### PR TITLE
Allow more general objectives in combinatorial layers

### PR DESCRIPTION
**Pull request by BatyLeo**
See #25.

- [x] `PerturbedAdditive`
- [x] `PerturbedMultiplicative`
- [x] `SPOPLusLoss`
- [x] More tests
- [x] `RegularizedGeneric`
- [x] Adapt current tests to new framework
- ~`StructuredSVM`~
- ~`Interpolation`~

### CODE REVIEW

**Comment by BatyLeo** [comment]
Current state of this pull request:

### Perturbed (additive and multiplicative)
Seems to work, by imitation (Fenchel-Young loss) as well as by experience (PushForward)

### SPO+
Works, but should only really work with $\alpha = 2$.
Indeed, with gradient and loss values are wrong in theory when $\alpha\neq 2$. It could be easily fixed for $\alpha\neq 1$, but we would need to be able to compute $\arg\max_y \theta^\top g(y)$ for $\alpha = 1$.

File: src/fenchel_young/fenchel_young.jl [hunk 1]
```diff
@@ -54,6 +54,19 @@ function fenchel_young_loss_and_grad(
...
+function fenchel_young_loss_and_grad(
+    fyl::FenchelYoungLoss{P},
+    θ::AbstractArray{<:Real},
+    y_true::AbstractArray{<:Real};
+    kwargs...,
+) where {P<:AbstractPerturbed{<:GeneralizedMaximizer}}
+    (; predictor) = fyl
+    F, almost_ŷ = fenchel_young_F_and_first_part_of_grad(predictor, θ; kwargs...)
+    l = F - objective_value(predictor.maximizer, θ, y_true; kwargs...)
+    g = predictor.maximizer.g(almost_ŷ) - predictor.maximizer.g(y_true)
```
Comment (BatyLeo) [inline src/fenchel_young/fenchel_young.jl:66]: This is not true if $g$ is non-linear in $y$. This needs to be fixed.

**Comment by BatyLeo** [comment]
Now that #80 and #65 are merged, I think this PR is finally ready to be merged.
I just need some feedback about the following:
- I'm not sure to be satisfied with the current name of the new structure `GeneralizedMaximizer`, do you have any better suggestion?
- Currently, `GeneralizedMaximizer` is only compatible with `PerturbedAdditive`, `PerturbedMultiplicative`, `SPOPlusLoss`, and `FenchelYoungLoss`. Do you think we should also add (if possible) support for `Regularized`/`SSVM`/`Interpolation`? I would say no, I think it's enough for what we need this feature for.

**Comment by gdalle** [comment]
I think support for Regularized would be nice. Does it also work with the latest Perturbed revamp?

**Comment by BatyLeo** [comment]
> I think support for Regularized would be nice. Does it also work with the latest Perturbed revamp?

- The problem with `Regularized` is that I'm not sure if it's really compatible with `RegularizedFrankWolfe`, because since $f(\theta) = \arg\max_y \theta^\top g(y) + h(y)$ is not a linear solver we cannot use it in the Frank Wolfe algorithm. I could probably make it compatible with `AbstractRegularized` tho, even if the Frank Wolfe forward pass won't work as expected.
- It works with the latest `PerturbedAdditive` and `PerturbedMultiplicative`, and is really useful when combined with a `FenchelYoungLoss`. It probably also works with a `PerturbedOracle` but its useless since we can directly give any oracle as input.

**Comment by gdalle** [comment]
> I could probably make it compatible with AbstractRegularized tho, even if the Frank Wolfe forward pass won't work as expected.

Indeed FW will fail, but the whole idea of `AbstractRegularized` was to allow for custom solvers, so I think it's worth the adaptation

**Comment by BatyLeo** [comment]
This would need a way for getting the maximizer of an `AbstractRegularized`, should we add one method to the `AbstractRegularized` interface? It would only be used when using `GeneralizedMaximizer`, that's a bit weird

**Comment by gdalle** [comment]
> It would only be used when using GeneralizedMaximizer, that's a bit weird

Is it a method that's easy to code?

**Comment by BatyLeo** [comment]
Yes, it would only need the following interface:
```julia
function get_maximizer end
```
We just need to specify that it needs to be implemented only for `AbstractRegularized` supporting `GeneralizedMaximizer`

**Comment by gdalle** [comment]
Your call. Btw you can also use RequiredInterfaces.jl to test the interfaces, it's really easy and convenient

**Comment by BatyLeo** [comment]
I added a new abstract type `AbstractRegularizedGeneralizedMaximizer <: AbstractRegularized` with the additional needed interface using RequiredInterfaces.jl. Now `GeneralizedMaximizer` can be regularized!

**Review by LouisBouvier — APPROVED** [review]
Nice job, everything seems ok!

**Pull request merged**